### PR TITLE
Lumen ai outlines

### PR DESCRIPTION
**Pull request by ahuang11**
I added outlines, but I don't think I'm formatting the input messages correctly, and thus, it's able to run, but sometimes crash, sometimes experience the dropped reference issue, and sometimes just returns the original spec.

I think I much prefer instructor.

image

### CODE REVIEW

File: lumen/ai/agents.py [hunk 1]
```diff
@@ -161,14 +161,14 @@ class LumenBaseAgent(Agent):
     user = param.String(default='Lumen')
 
     def _render_lumen(self, component: Component, message: pn.chat.ChatMessage = None):
-        def _render_component(spec, active):
+        async def _render_component(spec, active):
```
Comment (ahuang11) [inline lumen/ai/agents.py:164]: This doesn't fix the dropping reference issue.

File: lumen/ai/llm.py [hunk 2]
```diff
@@ -169,40 +181,83 @@ class OpenAI(Llm):
...
+        system: str = "",
+        **kwargs,
+    ):
+        from outlines import generate
+
+        generator = generate.json(create, kwargs.pop("response_model"))
+        message = "\n\n".join(
+            [f"System: {system}"]
+            + [f"{message['role']}: {message['content']}" for message in messages],
+        )
```
Comment (ahuang11) [inline lumen/ai/llm.py:250]: It doesn't accept the OpenAI format; I don't think this is how I should be formatting it.

Comment (ahuang11) [reply lumen/ai/llm.py:250]: I think I should be applying something like https://huggingface.co/blog/chat-templates but I need to load transformers too...

**Comment by philippjfr** [comment]
Closing for now. Can revive later if needed.

**Pull request closed without merging**